src/python/pants/bin/pants_loader.py

```diff
@@ -101,9 +102,9 @@ def main(cls) -> None:
         cls.ensure_locale()
 
         sys.setrecursionlimit(int(os.environ.get(RECURSION_LIMIT, "10000")))
+        os.environ.setdefault(WORK_DIR, os.getcwd())
```

Comment:
This is definitely not the way to do this.

So, right here:
https://github.com/pantsbuild/pants/blob/0107763505df49d428386e5222afd108fec27d2f/src/rust/engine/src/externs/interface.rs#L558-L566

There is an `exe.cmd.working_dir` `PathBuf` we are not passing but should be.

That call goes here:
https://github.com/pantsbuild/pants/blob/d68cc258e775db1522578a53d6b63870b9b45d34/src/python/pants/bin/daemon_pants_runner.py#L146-L174

And that is exactly and the only time we should need to set / derive CWD. For non-pantsd runs we should need to.

Reply:
OK, I think I'm slightly confused now ;)

> This is definitely not the way to do this.

As it would be unreliable or..? From the information that follows it is still not clear to me why it would be definitely wrong 😬 

> There is an exe.cmd.working_dir PathBuf we are not passing but should be.

Do you mean as an additional (new) argument to the `DaemonPantsRunner.__call__` method?

> And that is exactly and the only time we should need to set / derive CWD.

Do you mean, actually setting the process'es working directory here, or just capturing into the pants env var?

> For non-pantsd runs we should not need to set it.

~it being the CWD right?, we still need to capture CWD into the env var, as the CWD will be set to build root.~ Ah, for non-daemon runs, CWD is not changed so we can rely on that alone. 


Typing all this out, I get the feeling that half the remark is about the CWD handling being broken; is that it?  (after learning that non-pantsd runs behaves differently I think I get it)


So, you suggest scrap the custom env var and fix the CWD issue?

But, I'm not sure we won't break stuff if we run off of a different CWD than build root as there may be any number of assumptions about that.. although that would be the more correct approach in the long run.

Reply:
> But, I'm not sure we won't break stuff if we run off of a different CWD than build root as there may be any number of assumptions about that.. although that would be the more correct approach in the long run.

Makes me realize that this breaks non-pantsd runs not in build root unless we introduce a `os.chdir(build_root)` also for non-pantsd runs.

Reply:
I'm just pointing out the current env var set / read is hacky. It works, but it smells a bit patchy.

I posit that, for non-pantsd runs, no env var should be needed, the cwd is correct, say three levels deep in the repo. It's only for the same run under pantsd where cwd is incorrect. In that case pantsd is purposely demonized at the build root. But! In that exact and only case (the most common case to be sure), the nailgun protocol Pants uses to send the client request to the pantsd server already carries the client cwd as part of the protocol. We just neglected to use that. That's what I'm pointing out and suggesting you use, since it's purpose built for exactly this already. You may or may not still need an env var hack, but it should only be needed (set) exactly at the point the nail is being invoked and it should be set using the value already passed in the nailgun protocol (that the rust side of the nail neglects to pass to the Python side).

If the above paragraph is news, that's my point. You may still not want to use the existing mechanism, and that's fine I guess, but in that case the code setting the env var deserves a comment calling the shot - "We have the ng protocol for this but don't use it because...".

Reply:
Thank you for that clarification. Yes, ng is new territory for me. I think in the long run, we'd want to run with the CWD unchanged (i.e. be where ever the user was when invoking `pants`) but as an incremental step forward I'd prefer to keep CWD at build root and use the env var hack as work-around. Will call this out in comments.

Reply:
I linked this elsewhere, but the protocol (originally for JVM use) is defined here: http://www.martiansoftware.com/nailgun/protocol.html

I don't think many Pants devs know we use this or how the client / server flow works with pantsd.

Reply:
Oh ok.. :)

Reply:
so #18425 may be moot.

Reply:
Yes. There should be a follow up issue if the current hack lands, but I think it better to slow down and understand everything going on and just not introduce a hack from the get go. As a project with accelerating growth we have to try to fight becoming a patchy (did you know that's where that came from?!) server.

Reply:
To make this all concrete, though likely more confusing if you didn't know about this: 
https://github.com/pantsbuild/pants/blob/0b3f842cbc69d794d4bce8e2342c3455be0ff5b4/pants#L52

In other words there are two pants clients out there in the world. Your fix as it stands probably only works for the Python client and not this rust client (which is the future, but IIUC devs like @stuhood use daily).

Reply:
Yes, as you didn't suggest changing CWD, I'll take a moment to fix this properly. Also think I'll not change how non-pantsd runs would work, as if there would be any issues with CWD not being build root, that is new, work-aroundable and a bug that should be fixed. It also gives us a door in to easily check for any bad assumptions on the matter.

I think the follow up still stands too, in order to not rely on an env var is to actually change the forked process CWD to match that of the client (I may not have phrased that technically correct given my lacking understanding still of the daemon/nailgun interactions at play) but my idea is that we should have CWD mirror that of the user at all times, at which point we wouldn't need the `WORK_DIR` env var at all.

Or, I'll make that change here and now, and if it doesn't break--we're good already. 🤔

Reply:
When `pantsd` receives a connection, this code executes:
https://github.com/pantsbuild/pants/blob/d88e25829160d37aeefab4d7e470db52ccb95aac/src/rust/engine/nailgun/src/server.rs#L207-L263

On the stack there is a `nails::execution::Command`, which contains the working_dir of the client:
https://github.com/stuhood/nails/blob/67188138ce5b2a4855e2c4fe941e2ca519fa75eb/nails/src/execution.rs#L19-L25

So to get access to the client's working dir, you'd want to modify
https://github.com/pantsbuild/pants/blob/d88e25829160d37aeefab4d7e470db52ccb95aac/src/rust/engine/src/externs/interface.rs#L553-L581
...to pass the `exe.cmd.working_dir` through to the `runner` callable, and adjust the signature...
* here: https://github.com/pantsbuild/pants/blob/d88e25829160d37aeefab4d7e470db52ccb95aac/src/python/pants/engine/internals/native_engine.pyi#L243-L253
* and here: https://github.com/pantsbuild/pants/blob/d88e25829160d37aeefab4d7e470db52ccb95aac/src/python/pants/bin/daemon_pants_runner.py#L146-L174


Then, in `daemon_pants_runner.py`, pass the working_dir into other code.